compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/VectorizeIREEVectorExtOps.cpp

```diff
@@ -515,19 +525,29 @@ vectorizeLinalgExtGatherToTransferGather(RewriterBase &rewriter,
   ShapedType gatherTy = gatherOp.getOutputType();
   ShapedType sourceTy = gatherOp.getSourceType();
 
-  auto gatherVectorTy =
-      VectorType::get(gatherTy.getShape(), gatherTy.getElementType());
+  if (vectorSizes.empty()) {
+    vectorSizes = gatherTy.getShape();
+  }
+
+  auto gatherVectorTy = VectorType::get(vectorSizes, gatherTy.getElementType());
   // Rank-reduced to remove the innermost unit dim.
-  auto indicesVecTy =
-      VectorType::get(indicesTy.getShape().take_front(gatherOp.getBatchRank()),
-                      rewriter.getIndexType());
+  auto indicesVecTy = VectorType::get(
+      vectorSizes.take_front(gatherOp.getBatchRank()), rewriter.getIndexType());
 
-  // Read `indices` tensor via `vector.transfer_read` and cast from int to
-  // index.
+  // Read `indices` tensor via `vector.transfer_read`.
   Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
-  Value indicesVec = rewriter.create<vector::TransferReadOp>(
+  auto indicesVecRead = rewriter.create<vector::TransferReadOp>(
       loc, indicesVecTy.clone(indicesTy.getElementType()),
       gatherOp.getIndices(), SmallVector<Value>(indicesTy.getRank(), zero));
+  // Mask the index_vec read.
+  VectorType indicesMaskType = indicesVecTy.clone(rewriter.getI1Type());
+  SmallVector<OpFoldResult> gatherDims =
+      tensor::getMixedSizes(rewriter, loc, gatherOp.getOutput());
+  Value indicesMask = rewriter.create<vector::CreateMaskOp>(
+      loc, indicesMaskType,
+      ArrayRef(gatherDims).take_front(gatherOp.getBatchRank()));
+  Value indicesVec = maskOperation(rewriter, indicesVecRead, indicesMask);
+  // int -> index cast.
   indicesVec =
       rewriter.create<arith::IndexCastOp>(loc, indicesVecTy, indicesVec);
```

Comment:
Optional nit: I'd either drop the comments (because the code is self-documented to me), or have a mega comment in the front.